devito/ir/support/basic.py

```diff
@@ -477,7 +477,11 @@ def is_increment(self):
     def is_carried(self, dim=None):
         """Return True if a dimension-carried dependence, False otherwise."""
         try:
-            return (self.distance > 0) if dim is None else self.cause == dim
+            if dim is None:
+                return self.distance > 0
+            else:
+                testdim = dim.parent if dim.is_Sub else dim
```

Comment:
I'm slightly concerned about the proliferation of ``is_Sub`` in the code base.

So here's my proposal to mitigate this issue, which may also impact #504 .

We add a new method to ``Dimension``, let's call it (better names welcome) ``_bound_dimensions``. Then we do the following:

```
class Dimension(...):
    ...
    @property
    def _bound_dimensions(self):
        return {self}

...
class DerivedDimension(...):
    ...
    @property
    def _bound_dimensions(self):
        return {self, self.parent}
```

Now we can change ``is_carried`` (and, likewise, ``is_independent`` a few lines below) as:

```
try:
    if dim is None:
        return self.distance > 0
    else:
        return any(i == self.cause for i in self._bound_dimensions)
except TypeError:
    ...
```

Note that this way we're checking *both* the parent and the derived dimensions, which, in general, **should** be the right thing to do (I hope I'm not missing anything).

We could exploit this new method in #504 too, when calculating the distance (we could now check both the distance from the parent and from self). 

A benefit of this approach would be simpler portability if we start adding other types of weird dimensions

Thoughts? This is really open for discussion

Reply:
what do u think about the above @tjb900 @mloubout ?

Reply:
Sorry!  I've been meaning to get back to you on this, but I want to understand it a bit more and code up what you suggested both here and in #504 first.

I, too, find the creeping growth of is_Sub concerning - many of these exceptions feel like things that could also apply in other cases.